databricks/koalas/indexing.py

```diff
@@ -400,16 +399,23 @@ def _select_rows(self, rows_sel):
                 # If slice is None - select everything, so nothing to do
                 return None, None
             elif len(self._internal.index_columns) == 1:
-                start = rows_sel.start
-                stop = rows_sel.stop
-
+                sdf = self._kdf_or_kser._internal.sdf
                 index_column = self._kdf_or_kser.index.to_series()
-                index_data_type = index_column.spark_type
+
+                # get natural order from '__natural_order__' from start to stop
+                # based on index_columns to keep natural order.
+                start = sdf.select(NATURAL_ORDER_COLUMN_NAME) \
+                           .where(index_column._scol == rows_sel.start) \
+                           .first()[0]
+                stop = sdf.select(NATURAL_ORDER_COLUMN_NAME) \
+                          .where(index_column._scol == rows_sel.stop) \
+                          .first()[0]
```

Comment:
I think you can use case-when and avoid launching a job execution here. For instance,

```python
natural_order = F.when(
    index_column._scol == rows_sel.stop, F.col(NATURAL_ORDER_COLUMN_NAME)
).otherwise(F.lit(None))
cond.append(natural_order.isNotNull() & sdf[NATURAL_ORDER_COLUMN_NAME] <= natural_order)
```

Not tested but I guess this works?

Reply:
thanks for the comment!

oops i just checked this comment after using another way.

how about now implementation because we have to catch `KeyError` when given index name is not included in actual index when its type is `StringType`?

for example, when type of index is numeric, it is not raise KeyError like below.

```python
>>> pdf
   month  year  sale
0      1  2012    55
1      4  2014    40
2      7  2013    84
3     10  2014    31

>>> pdf.loc[0:1000]
   month  year  sale
0      1  2012    55
1      4  2014    40
2      7  2013    84
3     10  2014    31
```

but when the type of index is string, it raises KeyError like below.

```python
>>> pdf
            max_speed  shield
cobra               1       2
viper               4       5
sidewinder          7       8

>>> pdf['cobra':'hello']
Traceback (most recent call last):
...
KeyError: 'hello'
```

Reply:
That's all good but problem is `first()` triggers each job and we should avoid it.
Currently, it looks going to launch 3 jobs. It's better to make the API in single pass.

Reply:
okay i see. let me take manage this. i'll try the way you recommended 👍

Reply:
also, please check date data type
I think it is same issue 

```python
>>> dates = pd.date_range('20130101', periods=6)
>>> pdf = pd.DataFrame(np.random.randn(6, 4), index=dates, columns=list('ABCD'))
>>> pdf
                   A         B         C         D
2013-01-01  0.322155 -0.851969 -1.501208 -0.660413
2013-01-02 -0.388889  0.508557 -1.215688  0.761596
2013-01-03 -1.698188 -0.006648  2.111597 -1.404037
2013-01-04 -0.850982  1.449845 -0.259124  0.096576
2013-01-05  0.147729  0.604907  0.414314  0.125392
2013-01-06  0.016597 -0.102719 -0.965090  0.823600
>>> kdf = ks.from_pandas(pdf)
>>> pdf.loc[dates[0]:dates[1], 'D']
2013-01-01   -0.660413
2013-01-02    0.761596
Freq: D, Name: D, dtype: float64
>>> kdf.loc[dates[0]:dates[1], 'D']
```

Reply:
yeah it also fixed with existing changes

```
>>> kdf.loc[dates[0]:dates[1], 'D']
2013-01-01   -1.033028
2013-01-02    0.763634
Name: D, dtype: float64
```

Reply:
@HyukjinKwon 

hmm... i tried to fix, but i think it is hard to avoid using `first()` since we have to know the start & stop number exactly as an integer to make proper condition.

for example, it doesn't work properly like below since `natural_order` is also condition, not a number. (it always return the result only including `start` or `stop` row itself, not as a range)

```
>>> natural_order
Column<b'CASE WHEN (__index_level_0__ = viper) THEN __natural_order__ ELSE NULL END'>
>>> sdf.show()
+-----------------+---------+------+-----------------+
|__index_level_0__|max_speed|shield|__natural_order__|
+-----------------+---------+------+-----------------+
|            cobra|        1|     2|      25769803776|
|            viper|        4|     5|      60129542144|
|       sidewinder|        7|     8|      94489280512|
+-----------------+---------+------+-----------------+

>>> sdf.where(sdf[NATURAL_ORDER_COLUMN_NAME] <= natural_order).show()
+-----------------+---------+------+-----------------+
|__index_level_0__|max_speed|shield|__natural_order__|
+-----------------+---------+------+-----------------+
|            viper|        4|     5|      60129542144|
+-----------------+---------+------+-----------------+
```